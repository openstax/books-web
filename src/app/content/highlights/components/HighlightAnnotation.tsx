import { HTMLTextAreaElement } from '@openstax/types/lib.dom';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import Button from '../../../components/Button';
import * as Styled from './ShowMyHighlightsStyles';

interface HighlightAnnotationProps {
  annotation: string;
  isEditing: boolean;
  onSave: (annotation: string) => void;
  onCancel: () => void;
}

// tslint:disable-next-line:variable-name
const HighlightAnnotation = (
  { annotation, isEditing, onSave, onCancel }: HighlightAnnotationProps
) => {
  const [anno, setAnno] = React.useState(annotation);

  return <Styled.HighlightNote>
    {isEditing
      ? <FormattedMessage id='i18n:highlighting:card:placeholder'>
        {(msg: string) => <textarea
          value={anno}
          placeholder={msg}
          autoFocus={true}
          onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) => {
            setAnno(e.target.value);
          }}
        />}
      </FormattedMessage>
      : <Styled.HighlightNote>
        <span>
          <FormattedMessage id='i18n:toolbar:highlights:popup:body:note:text'>
            {(msg: Element | string) => msg}
          </FormattedMessage>
        </span>
        {annotation}
      </Styled.HighlightNote>
      }
    {isEditing && <Styled.HighlightEditButtons>
      <FormattedMessage id='i18n:highlighting:button:save'>
        {(msg: Element | string) => <Button
          data-testid='save'
          data-analytics-label='save'
          size='medium'
          variant='primary'
          onClick={() => onSave(anno)}
        >{msg}</Button>}
      </FormattedMessage>
      <FormattedMessage id='i18n:highlighting:button:cancel'>
        {(msg: Element | string) => <Button
          size='medium'
          data-analytics-label='cancel'
          data-testid='cancel'
          onClick={onCancel}
        >{msg}</Button>}
      </FormattedMessage>
    </Styled.HighlightEditButtons>}
  </Styled.HighlightNote>;
};

export default HighlightAnnotation;
