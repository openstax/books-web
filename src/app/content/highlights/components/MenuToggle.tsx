import React from 'react';
import styled from 'styled-components';
import { EllipsisV } from 'styled-icons/fa-solid/EllipsisV';
import theme from '../../../theme';

// tslint:disable-next-line:variable-name
export const MenuIcon = styled(EllipsisV)`
  height: 2rem;
  width: 2rem;
  padding: 0.2rem;
  color: ${theme.color.neutral.darkest};
  user-select: none;

  &:hover {
    color: ${theme.color.secondary.lightGray.darkest};
  }
`;

/*
  this should be a button but Safari and firefox don't support focusing buttons
  https://bugs.webkit.org/show_bug.cgi?id=22261
  https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
*/
// tslint:disable-next-line:variable-name
const MenuToggle = styled(({className}) => <div tabIndex={0} className={className}><MenuIcon /></div>)`
  border: none;
  display: block;
`;

export default MenuToggle;
