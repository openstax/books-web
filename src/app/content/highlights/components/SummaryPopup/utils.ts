import { Highlight } from '@openstax/highlighter/dist/api';
import queryString from 'querystring';
import { Book } from '../../../types';
import { findArchiveTreeNodeById } from '../../../utils/archiveTreeUtils';
import { getBookPageUrlAndParams } from '../../../utils/urlUtils';

export const createHighlightLink = ({ sourceId, id, anchor: elementId }: Highlight, book?: Book) => {
  const page = book ? findArchiveTreeNodeById(book.tree, sourceId) : undefined;
  const data = { target: JSON.stringify({ id, type: 'highlight' }) };
  const target = queryString.stringify(data) + `#${elementId}`;
  return `${page && book ? getBookPageUrlAndParams(book, page).url : ''}?${target}`;
};
