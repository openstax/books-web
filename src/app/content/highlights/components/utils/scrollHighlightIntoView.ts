import { Highlight } from '@openstax/highlighter';
import { HTMLElement } from '@openstax/types/lib.dom';
import { scrollIntoView } from '../../../../domUtils';

export default (highlight: Highlight, element?: HTMLElement | null) => {
  const firstElement = highlight.elements[0] as HTMLElement;
  const lastElement = highlight.elements[highlight.elements.length - 1] as HTMLElement;
  const otherElements = [lastElement];
  if (element) {
    otherElements.push(element);
  }
  scrollIntoView(firstElement, otherElements);
};
