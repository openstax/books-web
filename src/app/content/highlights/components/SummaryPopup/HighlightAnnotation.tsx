import { HTMLTextAreaElement } from '@openstax/types/lib.dom';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled from 'styled-components';
import Button from '../../../../components/Button';
import { textRegularStyle } from '../../../../components/Typography';
import theme from '../../../../theme';
import { HighlightEditButtons } from './styles';

// tslint:disable-next-line:variable-name
const HighlightNote = styled.div`
  ${textRegularStyle}
  padding-top: 1.2rem;
  display: flex;

  span {
    margin: 0 0.8rem 0 0;
    overflow: visible;
  }

  textarea {
    ${textRegularStyle}
    flex: 1;
    letter-spacing: 0;
    line-height: 20px;
    color: ${theme.color.text.label};
    padding: 8px;
  }
`;

interface HighlightAnnotationProps {
  annotation: string;
  isEditing: boolean;
  onSave: (annotation: string) => void;
  onCancel: () => void;
}

// tslint:disable-next-line:variable-name
const HighlightAnnotation = (
  { annotation, isEditing, onSave, onCancel }: HighlightAnnotationProps
) => {
  const [anno, setAnno] = React.useState(annotation);

  return <HighlightNote>
    {isEditing
      ? <FormattedMessage id='i18n:highlighting:card:placeholder'>
        {(msg: string) => <textarea
          value={anno}
          placeholder={msg}
          autoFocus={true}
          onChange={(e: React.ChangeEvent<HTMLTextAreaElement>) => {
            setAnno(e.target.value);
          }}
        />}
      </FormattedMessage>
      : <HighlightNote>
        <span className='highlight-note-text'>
          <FormattedMessage id='i18n:toolbar:highlights:popup:body:note:text'>
            {(msg: Element | string) => msg}
          </FormattedMessage>
        </span>
        {annotation}
      </HighlightNote>
      }
    {isEditing && <HighlightEditButtons>
      <FormattedMessage id='i18n:highlighting:button:save'>
        {(msg: Element | string) => <Button
          data-testid='save'
          data-analytics-label='save'
          size='medium'
          variant='primary'
          onClick={() => onSave(anno)}
        >{msg}</Button>}
      </FormattedMessage>
      <FormattedMessage id='i18n:highlighting:button:cancel'>
        {(msg: Element | string) => <Button
          size='medium'
          data-analytics-label='cancel'
          data-testid='cancel'
          onClick={onCancel}
        >{msg}</Button>}
      </FormattedMessage>
    </HighlightEditButtons>}
  </HighlightNote>;
};

export default HighlightAnnotation;
