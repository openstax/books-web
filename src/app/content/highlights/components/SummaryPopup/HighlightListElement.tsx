import { Highlight, HighlightUpdateColorEnum } from '@openstax/highlighter/dist/api';
import React from 'react';
import { useDispatch } from 'react-redux';
import styled, { css } from 'styled-components';
import { bodyCopyRegularStyle } from '../../../../components/Typography';
import theme from '../../../../theme';
import { deleteHighlight, updateHighlight } from '../../actions';
import { highlightStyles } from '../../constants';
import { popupBodyPadding } from '../HighlightStyles';
import ContextMenu from './ContextMenu';
import HighlightAnnotation from './HighlightAnnotation';
import HighlightDeleteWrapper from './HighlightDeleteWrapper';

// tslint:disable-next-line:variable-name
const HighlightOuterWrapper = styled.div`
  position: relative;
  overflow: unset;

  :not(:last-child) {
    border-bottom: solid 0.2rem ${theme.color.neutral.darker};
  }

  background: ${theme.color.neutral.base};
`;

// tslint:disable-next-line:variable-name
const HighlightContent = styled.div`
  ${bodyCopyRegularStyle}
  overflow: visible;

  * {
    overflow: initial;
  }
`;

// tslint:disable-next-line:variable-name
export const HighlightContentWrapper = styled.div`
  padding: 1.2rem ${popupBodyPadding}rem;
  ${(props: {color: string}) => {
    const style = highlightStyles.find((search) => search.label === props.color);

    if (!style) {
      return null;
    }

    return css`
      border-left: solid 0.8rem ${style.focused};

      ${HighlightContent} {
        background-color: ${style.passive};
      }

      .highlight-note-text {
        color: ${style.focused};
      }
    `;
  }}
`;

interface HighlightListElementProps {
  highlight: Highlight;
  locationFilterId: string;
  pageId: string;
}

// tslint:disable-next-line:variable-name
const HighlightListElement = ({ highlight, locationFilterId, pageId }: HighlightListElementProps) => {
  const [isEditing, setIsEditing] = React.useState(false);
  const [isDeleting, setIsDeleting] = React.useState(false);
  const dispatch = useDispatch();

  const updateAnnotation = (
    annotation: string
  ) => {
    dispatch(updateHighlight({
      highlight: {annotation},
      id: highlight.id,
    }, {
      locationFilterId,
      pageId,
    }));
    setIsEditing(false);
  };

  const updateColor = (color: string) => {
    dispatch(updateHighlight({
      highlight: {color: color as HighlightUpdateColorEnum},
      id: highlight.id,
    }, {
      locationFilterId,
      pageId,
    }));
  };

  const confirmDelete = () => {
    dispatch(deleteHighlight(highlight.id, {
      locationFilterId,
      pageId,
    }));
  };

  return <HighlightOuterWrapper>
    <ContextMenu
      color={highlight.color}
      onDelete={() => setIsDeleting(true)}
      onEdit={() => setIsEditing(true)}
      onColorChange={updateColor}
    />
    <HighlightContentWrapper color={highlight.color}>
      <HighlightContent
        className='summary-highlight-content'
        dangerouslySetInnerHTML={{ __html: highlight.highlightedContent }}
      />
      {highlight.annotation ? (
        <HighlightAnnotation
          annotation={highlight.annotation}
          isEditing={isEditing}
          onSave={updateAnnotation}
          onCancel={() => setIsEditing(false)}
        />
      ) : null}
    </HighlightContentWrapper>
    {isDeleting && <HighlightDeleteWrapper
      onCancel={() => setIsDeleting(false)}
      onDelete={confirmDelete}
    />}
  </HighlightOuterWrapper>;
};

export default HighlightListElement;
