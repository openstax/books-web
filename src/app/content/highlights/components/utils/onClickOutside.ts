import { EventListenerOptions, HTMLElement, MouseEvent } from '@openstax/types/lib.dom';
import React from 'react';
import { assertDocument, assertWindow } from '../../../../utils';

const onClickOutside = (
  element: React.RefObject<HTMLElement>,
  isFocused: boolean,
  cb: (ev: MouseEvent) => void,
  eventOptions?: EventListenerOptions
) => () => {
  if (typeof document === 'undefined') {
    return () => null;
  }

  const ifOutside = (e: MouseEvent) => {
    if (!(e.target instanceof assertWindow().Element)) {
      return;
    }
    if (!element.current || element.current.contains(e.target)) {
      return;
    }
    cb(e);
  };

  if (isFocused) {
    document.addEventListener('click', ifOutside, eventOptions);
  }

  return () => assertDocument().removeEventListener('click', ifOutside, eventOptions);
};

export const useOnClickOutside = (
  element: React.RefObject<HTMLElement>,
  isEnabled: boolean,
  cb: (e: MouseEvent) => void,
  eventOptions?: EventListenerOptions
) => {
  React.useEffect(onClickOutside(element, isEnabled, cb, eventOptions), [isEnabled]);
};

export default onClickOutside;
