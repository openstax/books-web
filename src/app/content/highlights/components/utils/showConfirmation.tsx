import React from 'react';
import ReactDOM from 'react-dom';
import uuid from 'uuid/v4';
import TestContainer from '../../../../../test/TestContainer';
import { assertDocument, assertNotNull } from '../../../../utils';
import ConfirmationModal from '../ConfirmationModal';

export default async() => {
  const document = assertDocument();
  const domNode = document.createElement('div');

  domNode.id = `dialog-${uuid()}`;
  const root = assertNotNull(document.getElementById('root'), 'root element not found');
  root.insertAdjacentElement('afterend', domNode);

  const userChoice: boolean = await new Promise((resolve) => {
    const confirm = () => resolve(true);
    const deny = () => resolve(false);
    ReactDOM.render(
      <TestContainer>
        <ConfirmationModal deny={deny} confirm={confirm} />
      </TestContainer>,
      domNode
    );
  });

  ReactDOM.unmountComponentAtNode(domNode);
  assertNotNull(domNode.parentElement, 'parent element not found').removeChild(domNode);

  return userChoice;
};
