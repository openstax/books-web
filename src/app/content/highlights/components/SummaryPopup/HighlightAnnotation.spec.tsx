import React from 'react';
import renderer from 'react-test-renderer';
import createTestServices from '../../../../../test/createTestServices';
import createTestStore from '../../../../../test/createTestStore';
import { book, page } from '../../../../../test/mocks/archiveLoader';
import createMockHighlight from '../../../../../test/mocks/highlight';
import { mockCmsBook } from '../../../../../test/mocks/osWebLoader';
import TestContainer from '../../../../../test/TestContainer';
import { runHooksAsync } from '../../../../../test/utils';
import { Store } from '../../../../types';
import { receiveBook, receivePage } from '../../../actions';
import { formatBookData } from '../../../utils';
import { highlightLocationFilters } from '../../selectors';
import { HighlightData } from '../../types';
import { getHighlightLocationFilterForPage } from '../../utils';
import MenuToggle from '../MenuToggle';
import HighlightAnnotation from './HighlightAnnotation';
import HighlightListElement from './HighlightListElement';
import * as utils from './utils';

describe('HighlightDeleteWrapper', () => {
  let store: Store;

  beforeEach(() => {
    store = createTestStore();
  });

  it('match snapshot when editing is set to false', async() => {
    store.dispatch(receiveBook(book));

    const component = renderer.create(<TestContainer store={store}>
      <HighlightAnnotation
        annotation='Some annotation'
        isEditing={false}
        // tslint:disable-next-line: no-empty
        onSave={() => {}}
        // tslint:disable-next-line: no-empty
        onCancel={() => {}}
      />
    </TestContainer>);

    await runHooksAsync();

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('match snapshot when editing is set to true', async() => {
    store.dispatch(receiveBook(formatBookData(book, mockCmsBook)));

    const component = renderer.create(<TestContainer store={store}>
      <HighlightAnnotation
        annotation='Some annotation'
        isEditing={true}
        // tslint:disable-next-line: no-empty
        onSave={() => {}}
        // tslint:disable-next-line: no-empty
        onCancel={() => {}}
      />
    </TestContainer>);

    await runHooksAsync();

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('properly fire onCancel and onSave props', async() => {
    store.dispatch(receiveBook(formatBookData(book, mockCmsBook)));

    let savedText = '';
    let cancelClicked = false;
    const annotation = 'TEST';

    const component = renderer.create(<TestContainer store={store}>
      <HighlightAnnotation
        annotation={annotation}
        isEditing={true}
        onCancel={() => { cancelClicked = true; }}
        onSave={(text) => { savedText = text; }}
      />
    </TestContainer>);

    await runHooksAsync();

    renderer.act(() => { return; });

    renderer.act(() => {
      const textarea = component.root.findByProps({ value: annotation });
      textarea.props.onChange({
        preventDefault: jest.fn(),
        target: {value: 'newAnno'},
      });
    });

    renderer.act(() => {
      const saveButton = component.root.findByProps({ 'data-testid': 'save' });
      saveButton.props.onClick();
      const cancelButton = component.root.findByProps({ 'data-testid': 'cancel' });
      cancelButton.props.onClick();
    });

    expect(savedText).toEqual('newAnno');
    expect(cancelClicked).toEqual(true);
  });
});

describe('Highlight annotation', () => {
  let store: Store;
  let services: ReturnType<typeof createTestServices>;
  let highlight: ReturnType<typeof createMockHighlight>;

  beforeEach(() => {
    store = createTestStore();
    services = createTestServices();
    highlight = createMockHighlight('asdf');
  });

  it('edits annotation from MH pop up and track analytics', async() => {
    store.dispatch(receiveBook(formatBookData(book, mockCmsBook)));
    store.dispatch(receivePage({...page, references: []}));

    const locationFilters = highlightLocationFilters(store.getState());
    const location = getHighlightLocationFilterForPage(locationFilters, page);

    jest.spyOn(utils, 'createHighlightLink')
      .mockReturnValue('/link/to/highlight');

    const component = renderer.create(<TestContainer services={services} store={store}>
      <HighlightListElement
        highlight={highlight as unknown as HighlightData}
        locationFilterId={location!.id}
        pageId={page.id}
      />
    </TestContainer>);

    await runHooksAsync();

    const track = jest.spyOn(services.analytics.editAnnotation, 'track');

    renderer.act(() => {
      const dropdown = component.root.findByType(MenuToggle);
      dropdown.props.onClick();
    });

    renderer.act(() => {
      const editButton = component.root.findByProps({ 'data-testid': 'edit' });
      editButton.props.onClick();
    });

    renderer.act(() => {
      const textarea = component.root.findByType('textarea');
      textarea.props.onChange({
        preventDefault: jest.fn(),
        target: {value: 'newAnno'},
      });
    });

    renderer.act(() => {
      const saveButton = component.root.findByProps({ 'data-testid': 'save' });
      saveButton.props.onClick();
    });

    expect(track).toHaveBeenCalled();
  });
});
