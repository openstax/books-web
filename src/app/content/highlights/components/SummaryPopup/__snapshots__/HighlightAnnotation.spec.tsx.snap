// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HighlightDeleteWrapper match snapshot when editing is set to false 1`] = `
.c0 {
  font-family: Neue Helvetica W01;
  color: #424242;
  font-size: 1.6rem;
  line-height: 2.5rem;
  padding-top: 1.2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 span {
  margin: 0 0.8rem 0 0;
  overflow: visible;
}

<div
  className="c0"
>
  <span
    className="highlight-note-text"
  >
    Note:
  </span>
  Some annotation
</div>
`;

exports[`HighlightDeleteWrapper match snapshot when editing is set to true 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-family: Neue Helvetica W01;
  border-radius: 0.2rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  min-width: 12rem;
  font-size: 1.6rem;
  height: 4rem;
  padding: 0 3rem;
  color: #fff;
  background-color: #f36b32;
  font-weight: bold;
  border: none;
}

.c3:hover {
  background-color: #e96128;
}

.c3:active {
  background-color: #df571e;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  cursor: pointer;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-family: Neue Helvetica W01;
  border-radius: 0.2rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  white-space: nowrap;
  min-width: 12rem;
  font-size: 1.6rem;
  height: 4rem;
  padding: 0 3rem;
  color: #5e6062;
  background-color: #fff;
  border: 1px solid #d5d5d5;
  font-weight: regular;
}

.c4:hover {
  background-color: #fafafa;
}

.c4:active {
  background-color: #e5e5e5;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0 8px;
}

.c2 button:first-child {
  margin-right: 8px;
}

.c0 {
  font-family: Neue Helvetica W01;
  color: #424242;
  font-size: 1.6rem;
  line-height: 2.5rem;
  padding-top: 1.2rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 span {
  margin: 0 0.8rem 0 0;
  overflow: visible;
}

.c1 {
  font-family: Neue Helvetica W01;
  color: #424242;
  font-size: 1.6rem;
  line-height: 2.5rem;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  line-height: 20px;
  color: #6f6f6f;
  padding: 8px;
}

<div
  className="c0"
>
  <textarea
    autoFocus={true}
    className="c1"
    onChange={[Function]}
    placeholder="Add a note..."
    value="Some annotation"
  />
  <div
    className="c2"
  >
    <button
      className="c3"
      data-analytics-label="save"
      data-testid="save"
      onClick={[Function]}
    >
      Save
    </button>
    <button
      className="c4"
      data-analytics-label="cancel"
      data-testid="cancel"
      onClick={[Function]}
    >
      Cancel
    </button>
  </div>
</div>
`;
